### ISSUE TITLE
Column order lost when writing Avro topic to JDBC sink

### ISSUE THREAD

**Issue by Oduig** [confluentinc/kafka-connect-jdbc#1014]
I looked around but could not find anyone mentioning this issue, aside from this Google group:
https://groups.google.com/forum/#!topic/debezium/cxgV0e_fHNI

The problem seems to be that the JDBC sink uses a HashMap to store column names:
https://github.com/confluentinc/kafka-connect-jdbc/blob/master/src/main/java/io/confluent/connect/jdbc/sink/metadata/FieldsMetadata.java#L86

This means columns are shuffled around when the JDBC sink creates a table. It would make more sense for the order of columns to be strict, for the following reasons.

1. Having important columns first makes data easier to explore.
2. Using a fixed order makes the table generation process predictable and deterministic.
3. Column order is important for query performance when it comes to the primary key. Even when I specify in `pk.fields` that the order of my primary key should be A, B, they still come out as B, A in the created table.
4. There is no reason why an arbitrary order is better, when the order is clearly defined in the Avro schema.

If I'm missing some existing functionality or configuration setting please let me know!

**Comment by gharris1727** [comment]
Have you experienced this issue recently with the connector? I believe we resolved this in #787. If it's still arbitrarily ordering the fields, then this would be a regression.

What might be happening is that the JDBC connector is doing some intentional re-ordering of the columns to locate PK columns first, before non-PK columns. This order _should_ be deterministic, but may not match the order of the fields as they were in the upstream Avro schema.

**Comment by Oduig** [comment]
Thank you, I didn't see that one before. Let's dive in. I am using a Kafka + Connect + Schema registry cluster hosted on Aiven. The JDBC connector has version 6.3.0 which should put it well past the fixed version (5.x.x).

I am using Avro with Schema registry for both the key and the value. For brevity I'll describe my schema as `(fieldName: fieldtypes)`, where the nullable fields all have a default of `null` and the others have no defaults.

- The key is `(customer: string, transactionid: string)`
- The value is `(customer: string, transactionid: string, timestamp: double, importdate: [null, double], merchant: [null, string], chargebackid: [null, string], chargebackreason: [null, string], fraudreason: [null, string])`

I am using the Avro schema on both the producer side and in the JDBC sink. I have no further explicit declaration of columns aside from the Avro schema. Here is the DDL of the auto-created table.

```
chargebackid text,
	importdate double precision,
	fraudreason text,
	merchant text,
	chargebackreason text,
	transactionid text not null,
	customer text not null,
	timestamp double precision not null,
	constraint my_table_name key (transactionid, customer)
```

I cannot find any logical patterns here, except that non-null columns seem to come after nullable columns. Other than that, the strings are neither in schema order, nor reverse order, nor alphabetical order, and the primary key fields neither come first nor last.

If there is any further debugging/diagnosis I can do to help, please let me know!

**Comment by gharris1727** [comment]
We have not released a 6.3.0 of this connector, and never will. I don't think you're using our software. Perhaps you're running https://github.com/aiven/aiven-kafka-connect-jdbc/ which [doesn't have the patch applied](https://github.com/aiven/aiven-kafka-connect-jdbc/blob/master/src/main/java/io/aiven/connect/jdbc/sink/metadata/FieldsMetadata.java#L133)?

**Comment by Oduig** [comment]
Ahhh that makes sense. There is only one JDBC sink in the UI which is called simply "JDBC sink", I assumed this was the Confluent sink. But you're right, the one I am using is maintained by Aiven.

My mistake! Good to know that this has been fixed, thank you so much for your helpful replies. I now know where to take this question. Have a good one! :100:
